File: microservice_case2/microservices/prescription-service/models/prescription.go
package models

import (
	"time"

	"gorm.io/gorm"
)

type Prescription struct {
	ID                 uint           `gorm:"primaryKey" json:"id"`
	PatientID          uint           `gorm:"index" json:"patient_id"`
	DoctorID           uint           `gorm:"index" json:"doctor_id"`
	AppointmentID      *uint          `json:"appointment_id,omitempty"`
	MedicationName     string         `json:"medication_name"`
	Description        string         `json:"description"`
	Dosage             string         `json:"dosage"`
	Frequency          string         `json:"frequency"`
	Quantity           int            `json:"quantity"`
	DurationDays       int            `json:"duration_days"`
	Instructions       string         `json:"instructions"`
	SideEffectsWarning string         `json:"side_effects_warning"`
	Status             string         `json:"status"` // active, completed, cancelled
	PrescribedDate     time.Time      `json:"prescribed_date"`
	ExpirationDate     time.Time      `json:"expiration_date"`
	CreatedAt          time.Time      `json:"created_at"`
	UpdatedAt          time.Time      `json:"updated_at"`
	DeletedAt          gorm.DeletedAt `gorm:"index" json:"deleted_at,omitempty"`

	Orders []PrescriptionOrder `gorm:"foreignKey:PrescriptionID" json:"orders,omitempty"`
}

type PrescriptionOrder struct {
	ID             uint           `gorm:"primaryKey" json:"id"`
	PrescriptionID uint           `gorm:"index" json:"prescription_id"`
	PharmacyID     uint           `gorm:"index" json:"pharmacy_id"`
	Quantity       int            `json:"quantity"`
	Status         string         `json:"status"` // pending, dispensed, completed
	OrderDate      time.Time      `json:"order_date"`
	DispensedDate  *time.Time     `json:"dispensed_date,omitempty"`
	CreatedAt      time.Time      `json:"created_at"`
	UpdatedAt      time.Time      `json:"updated_at"`
	DeletedAt      gorm.DeletedAt `gorm:"index" json:"deleted_at,omitempty"`

	Prescription Prescription `gorm:"foreignKey:PrescriptionID" json:"prescription,omitempty"`
}

type CreatePrescriptionRequest struct {
	PatientID          uint   `json:"patient_id" binding:"required"`
	DoctorID           uint   `json:"doctor_id" binding:"required"`
	AppointmentID      *uint  `json:"appointment_id,omitempty"`
	MedicationName     string `json:"medication_name" binding:"required"`
	Description        string `json:"description"`
	Dosage             string `json:"dosage" binding:"required"`
	Frequency          string `json:"frequency" binding:"required"`
	Quantity           int    `json:"quantity" binding:"required"`
	DurationDays       int    `json:"duration_days" binding:"required"`
	Instructions       string `json:"instructions"`
	SideEffectsWarning string `json:"side_effects_warning"`
}

type CreatePrescriptionOrderRequest struct {
	PharmacyID uint `json:"pharmacy_id" binding:"required"`
	Quantity   int  `json:"quantity" binding:"required"`
}

File: microservice_case2/tampilan micro/app/Models/Prescription.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\SoftDeletes;

class Prescription extends Model
{
    use HasFactory, SoftDeletes;

    protected $fillable = [
        'patient_id',
        'doctor_id',
        'appointment_id',
        'medication_name',
        'description',
        'dosage',
        'frequency',
        'quantity',
        'duration_days',
        'instructions',
        'side_effects_warning',
        'status',
        'prescribed_date',
        'expiration_date',
    ];

    protected $casts = [
        'prescribed_date' => 'datetime',
        'expiration_date' => 'datetime',
    ];

    /**
     * Get the patient
     */
    public function patient(): BelongsTo
    {
        return $this->belongsTo(User::class, 'patient_id');
    }

    /**
     * Get the doctor
     */
    public function doctor(): BelongsTo
    {
        return $this->belongsTo(User::class, 'doctor_id');
    }

    /**
     * Get the appointment
     */
    public function appointment(): BelongsTo
    {
        return $this->belongsTo(Appointment::class);
    }

    /**
     * Get prescription orders
     */
    public function prescriptionOrders(): HasMany
    {
        return $this->hasMany(PrescriptionOrder::class);
    }

    /**
     * Scope for active prescriptions
     */
    public function scopeActive($query)
    {
        return $query->where('status', 'active')
            ->where('expiration_date', '>', now());
    }

    /**
     * Scope for expired prescriptions
     */
    public function scopeExpired($query)
    {
        return $query->where(function ($q) {
            $q->where('status', 'expired')
                ->orWhere('expiration_date', '<=', now());
        });
    }
}
